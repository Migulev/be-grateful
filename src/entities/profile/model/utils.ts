import { Profile } from './types'

const getProfileDisplayName = (profile: Profile) => {
  if (profile.userName) return profile.userName
  if (profile.userName === '') return profile.email
  if (profile.name) return profile.name
  return profile.email
}

export const getProfileLetters = (profile: Profile) => {
  const displayName = getProfileDisplayName(profile)

  const [a, b] = displayName.split('@')[0].split(/\.|\s|-|_/)

  if (!b) {
    return `${a[0]?.toUpperCase() ?? ''}${a[1]?.toUpperCase() ?? ''}`
  }

  return `${a[0]?.toUpperCase() ?? ''}${b[0]?.toUpperCase() ?? ''}`
}
