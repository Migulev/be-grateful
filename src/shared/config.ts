export const maxGratitudeTextLength = 200
