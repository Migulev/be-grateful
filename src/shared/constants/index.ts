export * from './routes'

export const ONE_HOUR_TIME = 1000 * 60 * 60
