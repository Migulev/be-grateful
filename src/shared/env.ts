import { createEnv } from '@t3-oss/env-core'
import { z } from 'zod'

export const env = createEnv({
  clientPrefix: 'VITE_',
  client: {
    VITE_SUPABASE_URL: z.string().url(),
    VITE_SUPABASE_ANON_KEY: z.string().min(1),
    VITE_SUPABASE_AVATAR_BASE_URL: z.string().url(),
  },
  runtimeEnv: import.meta.env,
})
