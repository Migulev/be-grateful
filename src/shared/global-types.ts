export type DurationTW = 0 | 150 | 200 | 300 | 500 | 700 | 1000

export type WithImageUrl<T> = T & {
  avatarUrl: string
  image?: HTMLImageElement
}
