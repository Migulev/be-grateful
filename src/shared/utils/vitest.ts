export const customTextMatcher =
  (text: string) => (_: unknown, element: Element | null) => {
    const hasText = (node: Element | null) => node?.textContent === text
    const nodeHasText = hasText(element)
    const childrenDontHaveText = Array.from(element?.children || []).every(
      child => !hasText(child),
    )
    return nodeHasText && childrenDontHaveText
  }
