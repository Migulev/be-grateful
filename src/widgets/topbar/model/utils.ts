import { Profile } from './type'

const getProfileDisplayName = (profile: Profile) => {
  return profile.name ? profile.name : profile.email
}

export const getProfileLetters = (profile: Profile) => {
  const displayName = getProfileDisplayName(profile)

  const [a, b] = displayName.split('@')[0].split(/\.|\s|-|_/)

  if (!b) {
    return `${a[0]?.toUpperCase() ?? ''}${a[1]?.toUpperCase() ?? ''}`
  }

  return `${a[0]?.toUpperCase() ?? ''}${b[0]?.toUpperCase() ?? ''}`
}
