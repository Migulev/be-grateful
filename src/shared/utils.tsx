import {
  Children,
  Context,
  createContext,
  createElement,
  isValidElement,
  ReactNode,
  RefObject,
  useContext,
  useEffect,
  useState,
} from 'react'

import { ClassValue, clsx } from 'clsx'
import { twMerge } from 'tailwind-merge'

import { ObjectWithImageUrl } from './config/global-types'

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export function ComposeChildren({ children }: { children: ReactNode }) {
  const array = Children.toArray(children)
  const last = array.pop()

  return (
    <>
      {array.reduceRight(
        (child, element) =>
          isValidElement(element)
            ? createElement(element.type, element.props, child)
            : child,
        last,
      )}
    </>
  )
}

export function generateRandomId(): string {
  return `id_${Math.random().toString(36).slice(2, 11)}_${Date.now().toString(36)}`
}

export function convertDurationTW(duration: number) {
  switch (duration) {
    case 150:
      return 'duration-150'
    case 200:
      return 'duration-200'
    case 300:
      return 'duration-300'
    case 500:
      return 'duration-500'
    case 700:
      return 'duration-700'
    case 1000:
      return 'duration-1000'
    default:
      return 'duration-0'
  }
}

export function formatDate(isoString: string) {
  const date = new Date(isoString)
  const day = String(date.getDate()).padStart(2, '0')
  const month = String(date.getMonth() + 1).padStart(2, '0')
  const year = date.getFullYear()
  return `${day}.${month}.${year}`
}

export const useResizeTextarea = (
  element: RefObject<HTMLTextAreaElement>,
  dependency: string,
) => {
  useEffect(() => {
    if (element.current) {
      element.current.style.height = 'auto'
      element.current.style.height = `${element.current.scrollHeight}px`
    }
  }, [dependency, element])
}

export function createStrictContext<T>() {
  return createContext<T | null>(null)
}

export function useStrictContext<T>(context: Context<T | null>) {
  const value = useContext(context)
  if (value === null) throw new Error('Strict context not passed')
  return value as T
}

export function useMediaQuery(query: string) {
  const [value, setValue] = useState(false)

  useEffect(() => {
    function onChange(event: MediaQueryListEvent) {
      setValue(event.matches)
    }

    const result = matchMedia(query)
    result.addEventListener('change', onChange)
    setValue(result.matches)

    return () => result.removeEventListener('change', onChange)
  }, [query])

  return value
}

// Define the hook type with a generic parameter extending HTMLElement
export const useOutsideClick = <T extends HTMLElement>(
  ref: React.RefObject<T> | React.RefObject<SVGSVGElement>,
  callback: () => void,
): void => {
  useEffect(() => {
    // Event listener to call the callback if clicked outside of the element
    const handleClickOutside = (event: MouseEvent) => {
      if (ref.current && !ref.current.contains(event.target as Node)) {
        callback()
      }
    }

    // Bind the event listener
    document.addEventListener('mousedown', handleClickOutside)

    // Remove the event listener on cleanup
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [ref, callback])
}

export function shuffleArray<T>(array: T[]): T[] {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[array[i], array[j]] = [array[j], array[i]]
  }
  return array
}

export function preloadImageInObject<T>(
  obj: ObjectWithImageUrl<T>,
): Promise<ObjectWithImageUrl<T>> {
  return new Promise((resolve, reject) => {
    const img = new Image()
    img.src = obj.avatarUrl
    img.onload = () => {
      obj.image = img
      resolve(obj)
    }
    img.onerror = err => reject(err)
  })
}

export function getRandomNumber(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min
}

export function wait(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export function convertToHsl(cssValue: string) {
  return 'hsl(' + cssValue + ')'
}
