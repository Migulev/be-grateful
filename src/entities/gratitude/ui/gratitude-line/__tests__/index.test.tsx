import { render, screen } from '@testing-library/react'
import { type Mock } from 'vitest'

import { customTextMatcher, useStrictContext } from '@/shared/utils'

import { GratitudeLine } from '../index'

vitest.mock('@/shared/utils', async importOriginal => {
  const actual = await importOriginal()
  if (typeof actual === 'object' && actual !== null) {
    return {
      ...actual,
      useStrictContext: vitest.fn(),
    }
  }
  return {
    useStrictContext: vitest.fn(),
  }
})

describe('GratitudeLine', () => {
  beforeEach(() => {
    ;(useStrictContext as Mock).mockReturnValue({ isTouchScreen: false })
  })

  it('renders the title correctly', () => {
    render(
      <GratitudeLine
        title="Test Title"
        onDelete={() => {}}
      />,
    )
    expect(
      screen.getByText(customTextMatcher('-  Test Title')),
    ).toBeInTheDocument()
  })

  it('applies optimistic styles when isOptimistic is true', () => {
    const { container } = render(
      <GratitudeLine
        title="Test Title"
        onDelete={() => {}}
        isOptimistic
      />,
    )
    expect(container.firstChild).toHaveClass('opacity-50')
  })

  it('applies duration-700 class when optimisticDuration is 700', () => {
    const { container } = render(
      <GratitudeLine
        title="Test Title"
        onDelete={() => {}}
        optimisticDuration={700}
      />,
    )
    expect(container.firstChild).toHaveClass('duration-700')
  })

  it('renders the BrowserViewDeleteButton', () => {
    render(
      <GratitudeLine
        title="Test Title"
        onDelete={() => {}}
      />,
    )
    expect(screen.getByTestId('browser-view-delete-button')).toBeInTheDocument()
  })

  it('renders the MobileViewDeleteButton', () => {
    ;(useStrictContext as Mock).mockReturnValue({ isTouchScreen: true })
    render(
      <GratitudeLine
        title="Test Title"
        onDelete={() => {}}
      />,
    )
    expect(screen.getByTestId('mobile-view-delete-button')).toBeInTheDocument()
  })
})
