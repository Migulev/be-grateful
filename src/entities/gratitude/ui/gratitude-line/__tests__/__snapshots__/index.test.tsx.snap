// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GratitudeLine > renders the BrowserViewDeleteButton 1`] = `
<div
  class="group/button absolute -left-5 flex h-full w-10 items-center justify-center"
  role="browser-view-delete-button"
>
  <svg
    class="lucide lucide-x size-5 cursor-pointer text-destructive opacity-0 group-hover/button:opacity-100"
    fill="none"
    height="24"
    role="x-delete-button"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M18 6 6 18"
    />
    <path
      d="m6 6 12 12"
    />
  </svg>
  <svg
    class="lucide lucide-ellipsis-vertical absolute size-5 cursor-pointer opacity-0 group-hover/button:invisible group-hover/line:opacity-50"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="12"
      r="1"
    />
    <circle
      cx="12"
      cy="5"
      r="1"
    />
    <circle
      cx="12"
      cy="19"
      r="1"
    />
  </svg>
</div>
`;

exports[`GratitudeLine > renders the MobileViewDeleteButton 1`] = `
<svg
  class="lucide lucide-ellipsis-vertical absolute -left-3 size-5 cursor-pointer opacity-50"
  fill="none"
  height="24"
  role="mobile-view-delete-button"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    cx="12"
    cy="12"
    r="1"
  />
  <circle
    cx="12"
    cy="5"
    r="1"
  />
  <circle
    cx="12"
    cy="19"
    r="1"
  />
</svg>
`;
